src/python/zquantum/core/wavefunction.py

```diff
@@ -0,0 +1,142 @@
+from math import log2
+from typing import Any, Dict, List, Union
+from warnings import warn
+
+import numpy as np
+from sympy import Abs, Matrix, Symbol
+
+
+def _cast_sympy_matrix_to_numpy(sympy_matrix, complex=False):
+    new_type = np.complex128 if complex else np.float64
+
+    try:
+        return np.array(sympy_matrix, dtype=new_type).flatten()
+    except TypeError:
+        return np.array(sympy_matrix, dtype=object).flatten()
+
+
+class Wavefunction:
+    """
+    A simple wavefunction data structure that can
+    be used to calculate amplitudes of quantum states.
+
+    Args:
+        amplitude_vector: the initial amplitudes of the system,
+            can either be a NumPy ndarray or a SymPy Matrix
+    """
+
+    def __init__(
+        self, amplitude_vector: Union[np.ndarray, Matrix, List[complex]]
```

Comment:
Why sympy Matrix and not Vector?

Reply:
Vectors in Sympy are designed for the purpose of spatial coordinates and not as a generic mathematical vector.

More can be seen [here](https://docs.sympy.org/latest/modules/vector/index.html).

Reply:
Good to know, thanks!